### PR TITLE
Add provisioning endpoint to get Slack channel info

### PR DESCRIPTION
**Pull request by jaller94**
A first idea on how to solve #569

### CODE REVIEW

**Comment by jaller94** [comment]
Judging by the CI, no. I'll take this on tomorrow.

**Comment by jaller94** [comment]
@Half-Shot This is ready for review.

File: src/Main.ts [hunk 1]
```diff
@@ -1120,6 +1120,74 @@ export class Main {
         }
     }
 
+    public async getChannelInfo(opts: {
+        slack_channel_id: string,
+        slack_bot_token: string,
+        team_id: string,
+    }): Promise<ConversationsInfoResponse|void> {
+        let slackClient: WebClient | undefined;
```
Comment (Half-Shot) [inline src/Main.ts:1128]: Nit but, the spacing here is a bit odd.

File: src/Main.ts [hunk 2]
```diff
@@ -1120,6 +1120,74 @@ export class Main {
...
+        let teamEntry: TeamEntry | null = null;
+        let teamId: string = opts.team_id;
+
+        const existingChannel = opts.slack_channel_id ? this.rooms.getBySlackChannelId(opts.slack_channel_id) : null;
+
+        if (existingChannel) {
+            throw Error("Channel is already bridged! Unbridge the channel first.");
+        }
+
+        if (opts.slack_bot_token) {
```
Comment (Half-Shot) [inline src/Main.ts:1138]: We shouldn't be passing the slack token in here

File: src/Main.ts [hunk 3]
```diff
@@ -1120,6 +1120,74 @@ export class Main {
...
+                teamId = await this.clientFactory.upsertTeamByToken(opts.slack_bot_token);
+                log.info(`Found ${teamId} for token`);
+            } catch (ex) {
+                log.error("Failed to action link because the token couldn't used:", ex);
+                throw Error("Token did not work, unable to get team");
+            }
+        }
+
+        // else, assume we have a teamId
+        if (teamId) {
```
Comment (Half-Shot) [inline src/Main.ts:1154]: We should always have a teamId. If the team doesn't exist yet, the calling function should make sure we do create it.

File: src/Main.ts [hunk 4]
```diff
@@ -1120,6 +1120,74 @@ export class Main {
...
+
+            teamEntry = await this.datastore.getTeam(teamId);
+            if (!teamEntry) {
+                throw Error("Team ID provided, but no team found in database");
+            }
+        }
+
+        let channelInfo: ConversationsInfoResponse | undefined;
+        if (slackClient && opts.slack_channel_id && opts.team_id) {
+            // PSA: Bots cannot join channels, they have a limited set of APIs https://api.slack.com/methods/bots.info
```
Comment (Half-Shot) [inline src/Main.ts:1170]: This comment is redundant. We aren't trying to join a channel for channel info

File: src/Main.ts [hunk 5]
```diff
@@ -1120,6 +1120,74 @@ export class Main {
         }
     }
 
+    public async getChannelInfo(opts: {
```
Comment (Half-Shot) [inline src/Main.ts:1123]: I'd drop the opts object here and have them as real parameters, it's a bit of an ugly design pattern.

File: src/Main.ts [hunk 6]
```diff
@@ -1120,6 +1120,74 @@ export class Main {
...
+            }
+
+            teamEntry = await this.datastore.getTeam(teamId);
+            if (!teamEntry) {
+                throw Error("Team ID provided, but no team found in database");
+            }
+        }
+
+        let channelInfo: ConversationsInfoResponse | undefined;
+        if (slackClient && opts.slack_channel_id && opts.team_id) {
```
Comment (Half-Shot) [inline src/Main.ts:1169]: As per above, we can assume a teamId and a channelId

File: src/Main.ts [hunk 7]
```diff
@@ -1120,6 +1120,74 @@ export class Main {
...
+            if (!channelInfo.ok) {
+                if (channelInfo.error === 'channel_not_found') {
+                    return;
+                }
+                log.error(`conversations.info for ${opts.slack_channel_id} errored:`, channelInfo.error);
+                throw Error("Failed to get channel info");
+            }
+        }
+
+        if (opts.slack_channel_id &&
```
Comment (Half-Shot) [inline src/Main.ts:1182]: We can assume a channelId at this point

File: src/Main.ts [hunk 8]
```diff
@@ -1120,6 +1120,74 @@ export class Main {
...
+            this.allowDenyList.allowSlackChannel(opts.slack_channel_id, channelInfo?.channel.name) !== DenyReason.ALLOWED) {
+            log.warn(`Channel ${opts.slack_channel_id} is not allowed to be bridged`);
```
Comment (Half-Shot) [inline src/Main.ts:1184]: Probably not a warning. Also not sure the error is helpful as we're not trying to bridge it.

File: src/Provisioning.ts [hunk 9]
```diff
@@ -282,6 +282,57 @@ export class Provisioner {
         });
     }
 
+    @command("user_id", "channel_id", "team_id", "bot_token")
```
Comment (Half-Shot) [inline src/Provisioning.ts:285]: We don't need the bot_token

File: src/Provisioning.ts [hunk 10]
```diff
@@ -282,6 +282,57 @@ export class Provisioner {
...
+        if (!(await this.main.matrixUserInSlackTeam(opts.team_id, userId))) {
+            return Promise.reject({
+                code: HTTP_CODES.FORBIDDEN,
+                text: `${userId} is not in this team.`,
+            });
+        }
+
+        let channelInfo;
+        try {
+            channelInfo = await this.main.getChannelInfo(opts);
```
Comment (Half-Shot) [inline src/Provisioning.ts:314]: You could simplyify this code a lot by having the res.json part inside the try catch statement, 
```suggestion
            const channelInfo = await this.main.getChannelInfo(opts);
            if (!channelInfo) {
                return res.status(HTTP_CODES.NOT_FOUND).json({
                    message: 'Slack channel not found',
                });
            }
            res.json({
                name: channelInfo.channel.name,
                memberCount: channelInfo.channel.num_members,
            });
```

**Review by Half-Shot — COMMENTED** [review]
Looks plausible but there is a lot of cruft to remove here.

File: src/Main.ts (hunk 2 above)
Comment (jaller94) [reply src/Main.ts]: Removed everything related to botToken for getChannelInfo.

File: src/Main.ts (hunk 1 above)
Comment (jaller94) [reply src/Main.ts]: Removed the spaces

File: src/Main.ts (hunk 5 above)
Comment (jaller94) [reply src/Main.ts]: Unwrapped the parameters.

File: src/Main.ts (hunk 3 above)
Comment (jaller94) [reply src/Main.ts]: The calling function now ensures it's a non-empty string.
Removed the condition, keeping its content.

File: src/Main.ts (hunk 6 above)
Comment (jaller94) [reply src/Main.ts]: Removed the two conditions.

File: src/Main.ts (hunk 4 above)
Comment (jaller94) [reply src/Main.ts:1144]: Removed the comment.

File: src/Main.ts (hunk 7 above)
Comment (jaller94) [reply src/Main.ts]: Removed the condition.

File: src/Main.ts (hunk 8 above)
Comment (jaller94) [reply src/Main.ts]: Replaced with an empty return value.
The API now says:
`{
    message: 'Slack channel not found or not allowed to be bridged',
}`

File: src/Provisioning.ts (hunk 9 above)
Comment (jaller94) [reply src/Provisioning.ts]: Removed

File: src/Provisioning.ts (hunk 10 above)
Comment (jaller94) [reply src/Provisioning.ts]: Applied suggestion.

File: src/Main.ts (hunk 8 above)
Comment (jaller94) [reply src/Main.ts]: Suggestion to differentiate between "not found" and "not allowed" because I don't think this is confidential.

**Review by Half-Shot — APPROVED** [review]

**Pull request merged**